frontend/src/stimulus/controllers/pattern-input.controller.ts

```diff
@@ -378,11 +378,11 @@ export default class PatternInputController extends Controller {
   }
 
   private tagInvalidTokens():void {
-    this.contentTarget.querySelectorAll('[data-role="token"]').forEach((element) => {
-      const token = element.textContent?.trim();
-
+    this.contentTarget.querySelectorAll('[data-role="token"]').forEach((element:HTMLElement) => {
       let exists = false;
-      this.validTokens.forEach((prop) => { if (prop === token) { exists = true; } });
+      Object.keys(this.validTokenMap).forEach((prop) => {
+        if (prop === element.dataset.prop) { exists = true; }
+      });
```

Comment:
It's not possible to break out of a `forEach` loop early:

```suggestion
let exists = Object.keys(this.validTokenMap).some((prop) => prop === element.dataset.prop);
```

Reply:
true, way better loop